#Problem 1(a).  Reading Amazon Reviews.

In this problem, we will analyze Amazon reviews to determine what characteristics make them most helpful.

Download the file of Amazon gourmet food reviews from the [Stanford Large Network Dataset Collection](https://snap.stanford.edu/data/web-FineFoods.html).   (Your computer may already have a utility installed that can unzip the archive as a text file; if not, [7-zip](http://www.7-zip.org/) is a useful utility for Windows. You can also do a web search for ```open .gz files online``` and use an online utility.)

Create a pandas DataFrame object with the following entries for each review:

* Product ID
* Number of people who voted this review helpful
* Total number of people who rated this review
* Rating of the product
* Text of the review

For the second and third of these, the information will be given in the file as ```1/5```, which would correspond to 1 vote for helpful out of 5 people who rated the review.


In [2]:
import re
import pandas as pd


prod = []
helpful = []
total = []
rev = []
txt = []

prod_search = re.compile(r'product/productId: (.*)')
helpful_search = re.compile(r'review/helpfulness: (.*)')
rev_search = re.compile(r'review/score: (.*)')
txt_search = re.compile(r'review/text: (.*)')

with open('foods.txt', 'r') as f:
    for i, line in enumerate(f):
        
        prod_match = prod_search.match(line)
        helpful_match = helpful_search.match(line)
        rev_match = rev_search.match(line)
        txt_match = txt_search.match(line)
        
        if prod_match:
            prod.append(prod_match.group(1))
        if helpful_match:
            h, t = helpful_match.group(1).split('/')
            helpful.append(h)
            total.append(t)
        if rev_match:
            rev.append(rev_match.group(1))
        if txt_match:
            txt.append(txt_match.group(1))

cols = ['prod', 'helpful', 'total', 'rev', 'txt']
df = pd.DataFrame({'prod': prod,
                  'helpful': helpful,
                  'total': total,
                  'rev': rev,
                  'txt': txt
                  }, columns=cols)      
print(df.head())

         prod helpful total  rev  \
0  B001E4KFG0       1     1  5.0   
1  B00813GRG4       0     0  1.0   
2  B000LQOCH0       1     1  4.0   
3  B000UA0QIQ       3     3  2.0   
4  B006K2ZZ7K       0     0  5.0   

                                                 txt  
0  I have bought several of the Vitality canned d...  
1  Product arrived labeled as Jumbo Salted Peanut...  
2  This is a confection that has been around a fe...  
3  If you are looking for the secret ingredient i...  
4  Great taffy at a great price.  There was a wid...  


#Problem 1(b).  Analyzing review text.

Add columns to your DataFrame for the length of a review, the number of exclamation points in a review, and the fraction of people who rated a review helpful. You should calculate the fraction who rated a review helpful using the two columns you made in 1a, and a ratio of 1 helpful rating out of 5 total ratings should be entered as 0.2, not the string ```1/5```. If no people voted on whether a problem was helpful, the corresponding entry in your percentage helpful column should be ```NaN```.

In [5]:
df['rev_len'] = df.txt.str.len()
df['excl'] = df.txt.str.count('!')
df.helpful = pd.to_numeric(df.helpful)
df.total = pd.to_numeric(df.total)
df['frac'] = df.helpful / df.total
df.rev = pd.to_numeric(df.rev)
print(df.head())

         prod  helpful  total  rev  \
0  B001E4KFG0        1      1  5.0   
1  B00813GRG4        0      0  1.0   
2  B000LQOCH0        1      1  4.0   
3  B000UA0QIQ        3      3  2.0   
4  B006K2ZZ7K        0      0  5.0   

                                                 txt  rev_len  excl  frac  
0  I have bought several of the Vitality canned d...      263     0   1.0  
1  Product arrived labeled as Jumbo Salted Peanut...      190     0   NaN  
2  This is a confection that has been around a fe...      509     0   1.0  
3  If you are looking for the secret ingredient i...      219     0   1.0  
4  Great taffy at a great price.  There was a wid...      140     0   NaN  


#Problem 1(c).  Summary statistics.

How many reviews are in the data set?  What is the average length of a review (in characters)?  What is the average rating?  What is the greatest number of exclamation marks used in a single review?  Use the pandas package to answer these questions, then summarize your results in a markdown cell.

In [6]:
df.describe()

C:\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


helpful         total            rev        rev_len  \
count  568454.000000  568454.00000  568454.000000  568454.000000   
mean        1.743817       2.22881       4.183199     436.222083   
std         7.636513       8.28974       1.310436     445.339741   
min         0.000000       0.00000       1.000000      12.000000   
25%         0.000000       0.00000       4.000000     179.000000   
50%         0.000000       1.00000       5.000000     302.000000   
75%         2.000000       2.00000       5.000000     527.000000   
max       866.000000     923.00000       5.000000   21409.000000   

                excl           frac  
count  568454.000000  298402.000000  
mean        0.758047       0.776975  
std         1.564842       0.346321  
min         0.000000       0.000000  
25%         0.000000            NaN  
50%         0.000000            NaN  
75%         1.000000            NaN  
max        84.000000       3.000000

The total review count is 568,454. The average length of a review is approximately 436 characters. The average rating is 4.183199. 

#Problem 1(d).  Export.

Save your DataFrame as a .csv file suitable for future analysis in R.  Your .csv file should not include the review text column, as the presence of commas and quotation marks will make reading the file difficult.  You should also convert entries from ```NaN``` to the empty string before saving.

In [7]:
df_s = df.ix[:,df.columns != 'txt']
df_s.fillna('')
df_s.to_csv('df.csv', index=False)